### ISSUE TITLE
Create `<DashboardSharingSettings />`

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#4822]
## Feature Description

This issue is for building out the main interface for dashboard sharing settings which will later be integrated with the dashboard sharing settings button + modal being introduced in #4820. Since the integration is happening at a later stage, these issues can happen in parallel using Storybook for the development environment.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new `DashboardSharingSettings` component should be created in `assets/js/components/dashboard-sharing`
* Should use components created in other issues:
	- #4821 
	- Sub-components may be created as needed
* The design should follow the relevant portions of the [Figma document](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=737%3A14290) for dashboard sharing
	- The component should include everything within the contents of the modal (not the modal itself)
	- Any deviation from the design should be noted in the QA brief
	- Attention should be paid to differences between phone/mobile and larger viewports
* The table of Products (modules) should be primarily sorted by modules that the user owns the connection for first, can manage the settings for second, and then by the module's `order` property
	- The end result should be a table where the modules the user has the most control over are at the top and least/no control over at the bottom
* Note the difference in interface between single-admin and multi-admin environments (see `hasMultipleAdmins` on the `core/site` store)
	- Single admin environments have columns for `Product`, `Who can view`
	- Multi- admin environments have columns for `Product`, `Who can view`, `Who can manage view access` (column headers may change in the future)
* For modules that a user cannot manage sharing for, the static message should be shown in place of the role select:  
	> Contact managing user to manage view access
	- For multi-admin environments, the select for a module's sharing management is replaced by additional information about the module's owner where `admin-2` is the user's login/username:  
	> Managed by **admin-2** ℹ️

	The ℹ️ indicates the icon + tooltip as shown in the design
* Choices for who can manage view access correspond to the options for a module's sharing management:
	- `owner` "Only me" (default)
	- `all_admins` "All Admins"
* Rows for [shared ownership modules](https://github.com/google/site-kit-wp/issues/4791) are special in that these are manageable by all admins – their sharing management setting (not role select) should be disabled
* Rows for other modules become disabled while editing the shared roles for a module
* Applying the sharing settings should work similar to module settings (except there is only an edit view rather than separate view + edit)
	- Submitting can only happen if settings have changed (see `haveSharingSettingsChanged()`)
	- Submitting settings using `saveSharingSettings()`
	- A spinner should display (not shown in design) while the submit action is in progress and all setting fields should be disabled
* When changes to user roles are made, an informative dialog should be shown in the footer of the settings to inform the user of the the implications of the change
	- A similar but different message is shown if changing a module's settings management to "All admins" which would take precedence over the message about a change in selected roles
* Storybook stories should be added for the component, no integration should be done yet
* URLs for any _Learn More_ links should use https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/

## Implementation Brief

- Create `DashboardSharingSettings` component in `assets/js/components/dashboard-sharing`.

### Grouping the modules by owned, manageable and the rest
- Get all the modules from `getModules()` selector.
- Return `null` if the `modules` is not available.
- Create `PERMISSION_MANAGE_MODULE_SHARING_OPTIONS` constant with the value `googlesitekit_manage_module_sharing_options` in `assets/js/googlesitekit/datastore/user/constants.js`.
- To get the owned and manageable modules, loop through the `modules` object and use the `hasCapability()` selector by passing the `PERMISSION_MANAGE_MODULE_SHARING_OPTIONS` permission and the module slug.
- Add `sharingManagement` property using the `getSharingManagement(moduleSlug: string)` selector which is being implemented in #4795.
- The rest of the other modules that have the least priority should be sorted by the `order` property. Note that the`getModules()` selector already returns modules sorted by their `order`.
 
### Creating the layout
- The designs should match the Figma designs for the larger and smaller screens
- Use `Grid` -> `Row` -> `Cell` and each cell having an appropriate `lgSize`, `mdSize`, and `smSize`.
- All the cells beneath the same column will use the same sizes.
- Loop through the manageable modules array to create a grid layout.
- Use the `<ModuleIcon />` component and pass the `slug` to render the module icon.
- Single admin environments have columns for `Product`, and `Who can view`.
- Multi-admin environments should have an additional column `Who can manage view access`.
- Show the multi-admin column if there are multiple admins using the `hasMultipleAdmins()` selector.
- For modules that a user can manage sharing for, render the `<UseRoleSelect />` component.
- When editing the roles, all the setting fields should be disabled.
- For modules that a user cannot manage sharing for, the static message should be shown in place of the role select:
  >Contact managing user to manage view access
- For multi-admin environments:
  * the select for a module's sharing management is replaced by additional information about the module's owner where `admin-2` is the user's login/username:
  >Managed by admin-2 ℹ️
  * Get the module owner's login/username from `modules[module].owner.login`.
  * Export the ℹ️ from the Figma design
  * Use the `<Tooltip />` from `material-ui/core` that should have the following text where `admin-2` is the user's login/username:
  >Admin 2 has connected this and given managing permissions to all admins. You can change who can view this on the dashboard.
  * Choices for who can manage view access correspond to the options for a module's sharing management.
  * `owner` "Only me" (default) and `all_admins` "All Admins" should be available in the `sharingManagement` property obtained using the `getSharingManagement(moduleSlug: string)` selector (as we mentioned in the `Grouping the modules`).
  
### Dashboard Sharing Settings Footer
- Create `DashboardSharingSettingsFooter` component in `assets/js/components/dashboard-sharing`.
- Extract these lines from `assets/js/components/dashboard-sharing/DashboardSharingSettingsButton.js` to the new component.
  https://github.com/google/site-kit-wp/blob/a2823b7c40c1e5cd5e398d5699b0a7c9a5ea28e1/assets/js/components/dashboard-sharing/DashboardSharingSettingsButton.js#L131-L139
- Enable the `Apply` button only if settings have changed. In otherwords, disable the button if the settings haven't changed.
- Check whether the sharing settings have changed using the `haveSharingSettingsChanged()` selector.
- Render the `<Spinner />` component while the submit action is in progress and all setting fields should be disabled
- On submitting the settings should dispatch the `saveSharingSettings()` action which is being implemented in #4794 
- Check how the `Confirm Changes` button work in `assets/js/components/settings/SettingsActiveModule/Footer.js`.
- To display the footer notice upon changes:
  * Create a class `googlesitekit-sharing-settings-notice` or something appropriate. The styles should match the Figma designs.
  * For user role changes the message should be:
  >By clicking Apply, you’re granting the selected roles view-only access to data from the Google services you’ve connected via your account.
  ---------
  * If there is a change in the sharing settings management, that notice should take precedence over the selected role changes (message).
  * For sharing settings management changes the message should be:
  >By clicking Apply, you are giving other admins permission to manage view-only access on your behalf to data from the Google services you selected “All Admins” for.

- Add stories for the new `<DashboardSharingSettings />` component.

### Test Coverage

- Add unit tests for the `<DashboardSharingSettings />` component that covers the outlined features in the AC. 

## QA Brief

- Note the `Footer` component isn't rendered within the `DashboardSharingSettings` component because the `DashboardSharingSettings` has to be children of `DialogContent` and the `Footer` has to be children of `DialogFooter`. Hence the `Footer` is rendered in the `DashboardSharingSettingsButton` component.

- Verify the component in Storybook ("Components -> DashboardSharingSettings -> * ") matches the designs in Figma. Only the component itself, not the modal and modal footer. 
- Verify the functionality of the component as per the AC.
- Verify the `DashboardSharingSettingsButton` component in Storybook ("Components -> DashboardSharingSettingsButton -> * ") renders the `DashboardSharingSettings` component and matches the designs in Figma.
- Verify the `trackEvent` as per this [comment](https://github.com/google/site-kit-wp/issues/4822#issuecomment-1137022510)

## Changelog entry

* Add settings modal to allow admins to configure Dashboard Sharing.

**Comment by felixarntz** [comment]
@aaemnnosttv ACs LGTM! Just one thing to iterate (this may apply to other issues as well, so please keep it in mind for other ACs): Per #4949 (which came out of a conversation on another issue), let's always use `https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/` for all dashboard sharing-related "Learn more" links for now. #4949 will later be implemented to replace all of these links as needed.

**Comment by hussain-t** [comment]
@tofumatt, we need to bump up the estimate for adding test coverage. However, we don’t have anything between 19 to 30; what should I do?

**Comment by tofumatt** [comment]
@hussain-t In that case, let's go with 30; the estimate is at the high end of things but better to err on that side of things.

**Comment by tofumatt** [comment]
Alright, a lot to go through here but I think this all looks good. IB ✅

**Comment by asvinb** [comment]
Working branch: `feature/4822-dashboard-sharing-settings`

**Comment by hussain-t** [comment]
As this ticket is blocking #5221, @eugene-manuilov and I discussed adding the `trackEvent` call in the `DashboardSharingSettings` component from #5221 to this ticket. Hence I'm removing #5221 from blocking.

Within file `assets/js/components/dashboard-sharing/DashboardSharingSettings.js`:

- Add a call to `trackEvent` to the management permission dropdown change handler.
  - Event name:
    - When the dropdown value is `all_admins`, the event name should be `change_management_all_admins`.
    - When the dropdown value is `owner`, the event name should be `change_management_owner`.
  - Event category: `${viewContext}_sharing`.
  - Event label: the module slug for which the management setting was changed.
- Add a call to `trackEvent` to the **Apply** button click handler.
  - Event name: `settings_confirm`.
  - Event category: `${viewContext}_sharing`.
- Add a call to `trackEvent` to the **Cancel** button click handler.
  - Event name: `settings_cancel`.
  - Event category: `${viewContext}_sharing`.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@tofumatt @techanvil  I haven't started to QA this yet, but while creating my test plan I came across a console error when clicking on the 'Add roles' link for any of the modules. 

`Uncaught TypeError: Invalid attempt to spread non-iterable instance. In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`

![image](https://user-images.githubusercontent.com/73545194/171426943-d15deb74-408a-4cde-99fd-297deb1c9d23.png)

Also, just quickly reading through the QAB, it mentions checking to Storybook, but at this stage, we should only be testing on InstaWP test sites.  With this in mind are there any particular things we should be testing? My initial thoughts are every aspect of the Dashboard sharing & permissions functionality needs testing.

**Comment by tofumatt** [comment]
> Also, just quickly reading through the QAB, it mentions checking to Storybook, but at this stage, we should only be testing on InstaWP test sites. With this in mind are there any particular things we should be testing? My initial thoughts are every aspect of the Dashboard sharing & permissions functionality needs testing.

That was part of the QA Brief: to check storybook stories. For what it's worth I've looked through those already and they look good.

In terms of actual functionality, yeah, testing all aspects of Dashboard Sharing for this one—verifying that the changes made in the sharing settings are made for users—would be good. 👍🏻 

----

In terms of the JS error, I'm taking a look now, I missed that one. 😓  I'll make a PR if a change is needed.

**Comment by tofumatt** [comment]
@wpdarren Can you walk me through the steps you took to get that JS error? I wasn't able to recreate it in development.

**Comment by tofumatt** [comment]
Nevermind, I just noticed the error only happens when no sharing settings are set and the sharing button is clicked. I have a PR ready for this 🙂

**Comment by wpdarren** [comment]
### QA Update: ❌ 

On top of the issues [reported above](https://github.com/google/site-kit-wp/issues/4822#issuecomment-1144970488) I have additional observations. 

9. Extending from issue 1, reported above, this also occurs with the selection of `Who can manage view access` dropdown. In the example here, for Search Console, I select `'All Admin'` and click on the apply button. When I click back into the dashboard sharing settings modal, the change has not been saved.  No errors in the console.  Screencast here

https://user-images.githubusercontent.com/73545194/171674683-3256bf19-c33f-4676-8e2f-0210457e0073.mp4

10.  Odd UX/UI on small screen sizes when you are adding roles on a site that has two admin users.  The 
`Who can manage view access` title appears at the top, but the dropdown appears underneath the user roles. 

![image](https://user-images.githubusercontent.com/73545194/171676892-678f619d-8d34-4a16-a52a-466a32c6d01d.png)

11. Odd UX/Ui on small screen sizes when you select `Who can manage view access` dropdown.  The layout shifts and the dropbox appears further down the page.  Possibly linked with issue 10. above. 

![ds-5](https://user-images.githubusercontent.com/73545194/171678228-ee0d0d9d-e1cd-491e-93d0-5c4bd85ddad1.gif)

12. Minor styling issue on the `Managed by` tooltip when you have two admin users managing dashboard sharing settings. It slightly overlaps the modal as per screenshot. 

![image](https://user-images.githubusercontent.com/73545194/171679699-f1c15fe6-40b8-444a-85cf-669d84f128ed.png)

13. Should the tooltips appear on small screen sizes like mobiles? I am assuming that they would not function on mobile.

![image](https://user-images.githubusercontent.com/73545194/171680008-90bc4421-03da-4661-b292-e16dfe9251ec.png)

**Comment by wpdarren** [comment]
QA Update:

**Comment by asvinb** [comment]
@wpdarren Kindly find my comments pertaining to the points you raised:
* `2.` This is as per designs where no external icons is present. However for the sake of consistency, I think we should add it. There is a follow up PR for that.
* `4.` This has been implemented.
* `5.` I think there is some inconsistency in the Figma designs itself. If you look at this screen, https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=1214%3A18915, you will notice the font family used is `Roboto`, with a font weight of `500`. In screens where `Google Sans` is being used, the font weight is `400` but it does not look good since the font must be bolded. Furthermore, to be consistent with table headers, I think we can keep the font family to `Roboto` and the font weight to `500`.
* `6.` Same as above.
* `7.` Fixed.
* `8.` So we have 2 states for the button. One is the hover state and the other one is the focus states which are different. So when the dialog is closed, the button gets the focus back, which has a background color which is different from the hover state. This is expected.
* `10.` Maybe we need to create a followup issue for that since we don't have any designs on how it should look. I think to maximise the use of real estate screen, it's better to have the user roles to be 100% wide and then the dropdown goes below the roles.
* `11.` See above.
* `12.` The tooltips is managed outside of the modal, I don't think it's an issue to have part of it displayed outside of the modal.
* `13.` The tooltips should work on small screens. If you touch the icon for a small amount of time, the tooltip will be displayed.

**Comment by hussain-t** [comment]
@wpdarren, I have fixed `#3` and `#4` in this follow-up [PR](https://github.com/google/site-kit-wp/pull/5319). As for `#1` and `#9`, this is a backend issue that only the modules available in the `_googlesitekitDashboardSharingData.settings` are changeable.  @jimmymadon should be able to provide more context on this as he worked on the backed part of it.

**Comment by hussain-t** [comment]
@tofumatt, as per the AC, the shared ownership modules should have `All Admins` and be disabled:
>Rows for https://github.com/google/site-kit-wp/issues/4791 are special in that these are manageable by all admins – their sharing management setting (not role select) should be disabled

@wpdarren understood this point later and updated his post:
>This might be linked with issue 3 above. While logged into an admin account, for PSI and Idea Hub module they are managed by another admin user anadmin. When I login as anadmin I am unable to change Who can manage view access as they are inactive dropdowns. See screenshots below Is this expected?
Update: I just spotted this in the AC: Rows for https://github.com/google/site-kit-wp/issues/4791 are special in that these are manageable by all admins – their sharing management setting (not role select) should be disabled

>So, why show these two modules are managed by anadmin when they are managed by all admins?

So, `So, why show these two modules are managed by anadmin when they are managed by all admins?` is valid issue and I have fixed it in my [PR](https://github.com/google/site-kit-wp/pull/5319).

As for `#1` and `#9` that the roles aren't set after clicking `Apply`, as I mentioned in this [comment](https://github.com/google/site-kit-wp/issues/4822#issuecomment-1147640654) that this is a backend issue. The module has to be registered in the database for changing the sharing settings. @jimmymadon, could you shed more light on this as you worked on the backend. Thanks!

**Comment by wpdarren** [comment]
@hussain-t  @tofumatt re. 

> As for #1 and #9 that the roles aren't set after clicking Apply, as I mentioned in this https://github.com/google/site-kit-wp/issues/4822#issuecomment-1147640654 that this is a backend issue. The module has to be registered in the database for changing the sharing settings. @jimmymadon, could you shed more light on this as you worked on the backend. Thanks!

I am conscious that these are E2E blocking issues, and IMO, this ticket should not be moved to CR/MR and QA until they are fixed as they will need another PR.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

Can confirm that issue 1 and 9 are now fixed 🥳 

![image](https://user-images.githubusercontent.com/73545194/172604231-7a97e739-6aae-4c66-96ad-19471060f608.png)

**Verified:**

- The design follows the relevant portions of the [Figma document](https://www.figma.com/file/i04x83uA9U8HHBANlhS7Db/Dashboard-Sharing?node-id=737%3A14290) for dashboard sharing. 
- Tested on small screen sizes and larger viewports
- The table is sorted by the modules the user has the most control over are at the top and least/no control over at the bottom
- The Single admin environments have columns for Product, Who can view
- Multi- admin environments have columns for Product, Who can view, Who can manage view access
- Checked that the admin and non admin users could see the appropriate data and widgets to check. 

Managing user to manage view access

- For multi-admin environments, the select for a module's sharing management is replaced by additional information about the module's owner where admin-2 is the user's login/username: Managed by admin-2 with a tooltip.
- Submitting only happens if settings have changed 
- A spinner displays  while the submit action is in progress and all setting fields should be disabled
- When changes to user roles are made, an informative dialog is shown in the footer of the settings to inform the user of the the implications of the change. 
- A similar but different message is shown if changing a module's settings management to "All admins" which takes precedence over the message about a change in selected roles.
- URLs for any Learn More links use https://sitekit.withgoogle.com/documentation/using-site-kit/dashboard-sharing/

<details>
 <summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/73545194/172400467-098ffa59-2f96-4f8e-b2f5-e32c19142f81.png)
![image](https://user-images.githubusercontent.com/73545194/172400499-497d3bca-4691-40ef-af82-ef29313e5cd0.png)
![image](https://user-images.githubusercontent.com/73545194/172400545-3b6511da-f242-4b23-b7c9-98444681960b.png)
![image](https://user-images.githubusercontent.com/73545194/172400618-3a8caa37-66fb-46d6-bfce-39c8f96280f2.png)
![image](https://user-images.githubusercontent.com/73545194/172400825-e3073f8f-0758-4c7b-b3fa-f087c3b8a34c.png)

</details>
